Lambda-ify enabling GuardDuty

nancytao opened the pull request:
* Includes CloudFormation template to deploy stack 
* Lambda version is triggered by SNS and runs on a single account, specified by the SNS message
* Includes make targets to deploy, update, and test code

nancytao committed:
Lambda-ify enabling GuardDuty

jchrisfarris commented on lambda/enable_guardduty.py:
```diff
@@ -0,0 +1,330 @@
+    creds = get_creds(role_arn)
+    if creds is False:
+        logger.error(f"Unable to assume role {role_arn} to get parent organization")
+        return
```
This should also raise an error and cause the Lambda execution to error. return() ends up as a normal execution.

jchrisfarris commented on lambda/enable_guardduty.py:
```diff
@@ -0,0 +1,330 @@
+def process_region(event, region):
+    logger.info(f"Processing Region: {region}")
+
+    role_arn = create_role_arn(event["payer_account_id"], os.environ["PAYER_ROLE"])
```
What you're doing here assume that the Payer Account is also the GuardDuty master account. This is not always the case. For us, we have a different payers and only one GuardDuty account. This Lambda should execute in the GuardDuty/Security Account.

jchrisfarris commented on lambda/enable_guardduty.py:
```diff
@@ -0,0 +1,330 @@
+        raise KeyError(error_message)
+
+    # get parent organization's account_id
+    role_arn = create_role_arn(message['account_id'], os.environ['ASSUME_ROLE'])
```
Lets rename this to AUDIT_ROLE, and the other role as ACCEPT_ROLE

jchrisfarris commented on lambda/enable_guardduty.py:
```diff
@@ -0,0 +1,330 @@
+    gd_client = boto3.client(
+        'guardduty',
+        region_name=region,
+        aws_access_key_id=creds['AccessKeyId'],
```
Remove the creds, gd_client should be running against the GuardDuty Master which is what is executing this lambda.

jchrisfarris commented on lambda/enable_guardduty.py:
```diff
@@ -0,0 +1,330 @@
+        response = gd_client.list_detectors()
+        if len(response['DetectorIds']) == 0:
+            # We better create one
+            detector_id = create_parent_detector(gd_client, event, region)
```
If I had named this create_masteraccount_detector() it would have been less confusing

jchrisfarris commented on lambda/enable_guardduty.py:
```diff
@@ -0,0 +1,330 @@
+
+    gd_status = get_all_members(region, gd_client, detector_id)
+
+    account = describe_account(event)
```
This is where we get the root email for the target account. Good.

jchrisfarris commented on lambda/enable_guardduty.py:
```diff
@@ -0,0 +1,330 @@
+        if "accept_only" not in event or not event["accept_only"]:
+            invite_account(account, detector_id, gd_client, event, region)
+            time.sleep(3)
+        accept_invite(account, os.environ['ASSUME_ROLE'], event, region)
```
So this is the ACCEPT_ROLE here.

jchrisfarris commented on lambda/enable_guardduty.py:
```diff
@@ -0,0 +1,330 @@
+    if creds is False:
+        logger.error(
+            f"Unable to assume role into {account['Name']}({account['Id']}) to accept invite")
+        return(False)
```
We should raise an alert if we can't assume the role.

jchrisfarris replied:
s/alert/exception/

jchrisfarris commented on lambda/enable_guardduty.py:
```diff
@@ -0,0 +1,330 @@
+
+
+def create_role_arn(account_id, role_name):
+    return f"arn:aws:iam::{account_id}:role/{role_name}"
```
TIL a better way to "some{}string".format()

jchrisfarris commented on lambda/enable_guardduty.py:
```diff
@@ -0,0 +1,330 @@
+        return(False)
+
+
+def describe_account(event):
```
If we called this from the handler vs process_region() the lambda might run marginally faster

jchrisfarris reviewed (changes requested):
This looks good. The only thing needing fixing is how gd_client is created on line 81

nancytao committed:
Errors raised when fail to assume role, GuardDuty runs in GuardDuty Master, rename roles, other fixes

nancytao commented:
* Changed `get_creds()` to raise an error if assume role fails, instead of at the call-site 
* Renamed PAYER_ROLE -> AUDIT_ROLE and ASSUME_ROLE -> ACCEPT_ROLE 
* Changed GuardDuty to run in the GuardDuty master instead of in payer account (and added those permissions to CloudFormation) 
* Small refactor to main handler, moved `describe_account()` to main handler

Commit:
Disable the email notification of GuardDuty
